### ISSUE TITLE
Slider callback

### ISSUE THREAD

**Issue by kbssa** [s00500/ESPUI#127]
Every time I cahnge the slider value the callback is executed 3 times.

```
void slider(Control* sender, int type) {
	Serial.print("Slider: ID: ");
	Serial.print(sender->id);
	Serial.print(", Value: ");
	Serial.println(sender->value);
	return;
}

```
Slider: ID: 17, Value: 24
Slider: ID: 17, Value: 24
Slider: ID: 17, Value: 24
Slider: ID: 17, Value: 17
Slider: ID: 17, Value: 17
Slider: ID: 17, Value: 17
Slider: ID: 17, Value: 39
Slider: ID: 17, Value: 39
Slider: ID: 17, Value: 39
Slider: ID: 17, Value: 60
Slider: ID: 17, Value: 60
Slider: ID: 17, Value: 60

Is that normal to behave like that ?

**Comment by codefaux** [comment]
For anyone who runs into this, I just added a ratelimiter to my slider updates as a workaround. It's a generic ratelimiter, lightweight and clean, requires one 32-bit variable, one function call, a few cycles of math, etc.


`
#define SLIDER_UPDATE_TIME 100   //wait at least 100 ms before allowing another slider update
void ui_callback_slider(Control *sender, int type)
{
  static unsigned long last_slider_update = 0; // Track the time of the last slider update
 
  if ((type==SL_VALUE) && (millis() - last_slider_update >= SLIDER_UPDATE_TIME))
  {
    // DO THINGS

    last_slider_update = millis();
  }
  return;
}`

**Comment by iangray001** [comment]
I have been delving into this and it is far worse. **For** every slider, callbacks are added **to each** slider. This problem compounds with every slider in the UI. With 12 sliders you get a chain of around **77** callbacks. Even with continuous mode turned off, this is utterly destroying my target device. codefaux's rate limiter above wouldn't help because `malloc` is failing inside the JSON parsing before it even gets the chance to execute the callback to throw away the 77th identical update. 

The offending code is in `controls.js:rangeSlider()`:

```
var slider = $(".range-slider"),
  range = $(".range-slider__range"),
  value = $(".range-slider__value");

slider.each(function () {
  //...
range.on({ ... 
```

For each slider, the entire set of callbacks is added **to every** `range`. A hacky fix could be something like this:

```
var rangeSlider = function (isDiscrete) {
  var slider = $(".range-slider"),
    range = $(".range-slider__range"),
    value = $(".range-slider__value");

  slider.each(function () {
    value.each(function () {
      var value = $(this).prev().attr("value");
      $(this).html(value);
    });
  });

  range.each(function() {
    if($(this).attr("callbackSet") != "true") {

      if (!isDiscrete) {
        $(this).on({
          input: function () {
            sliderchange($(this).attr("id").replace(/^\D+/g, ""));
          },
        });
      } else {
        $(this).on({
          input: function () {
            $(this).next().html(this.value);
          },
          change: function () {
            sliderchange($(this).attr("id").replace(/^\D+/g, ""));
          },
        });
      }

      $(this).attr("callbackSet", "true");
    }
  });
};
```

Better would be to only call this at the end once all sliders have been processed, but this is a minimal fix.

Unfortunately the sliders are also broken if ranges other than 0-100 are used. I'm working on that as well!

**Comment by codefaux** [comment]
Woof -- you are absolutely correct, I didn't catch that myself. I'm early in prototyping a UI, so at the time I implemented the fix I only had four or five sliders. I hadn't caught how many times it was rate-limiting yet, and I hadn't set up a full target with the dozen or more I need.

I'm on hiatus for a bit but if there isn't a solid answer by the time I get around to it, I'll take a look as well. I'm not excited about fiddling with dense js, so good luck to you! Hah.

If anyone else winds up here -- My "workaround" of applying a rate-limiter will only hold for a few controls on a relatively low-traffic device.

**Comment by iangray001** [comment]
Right, I've added a PR to fix this callback issue. 

I've also debugged the issue of the sliders being broken for ranges other than 0-100 and it is a wonderful illustration of [default function arguments are the devil](https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/). There is a pretty severe bug in `ESPUIClass::addControl`. ESPUI.cpp:593 is the offending line:

`Control* control = new Control(type, label, callback, value, color, parentControl);`

Min and Max parameters are implemented as child controls of type `Min` and `Max` of the slider control, and they are added to the slider with the `addControl` function. However this fails. Why? If we check the prototype of the constructor:

`Control(ControlType type, const char* label, void (*callback)(Control*, int), const String& value,
        ControlColor color, bool visible = true, uint16_t parentControl = Control::noParent)`

Spot it? I didn't for ages! When `visible` was later added to the constructor in some feature update, because it and `parentControl` both have defaults then `addControl`'s now erroneous instantiation is silently papered over and every single component ends up with a parent of 65535. The fix is of course just to add a `true` back in for the `visible` parameter and it all works again. 

For future reference, this is why using multiple overloaded constructors is better because I would have immediately seen that my IDE was binding to one constructor not the other. 

I'll fold this all into a PR as well once I have figured out issue #136 which I suspect is related somehow...
